rids/rids.py

```diff
@@ -15,113 +15,111 @@
 # limitations under the License.
 
 
+
 """
 Main entry point for RIDS, a Remote Intrusion Detection System.
 """
 
-
 import ipaddress
 import json
-import os
-import subprocess
-import sys
-import urllib.request
+import logging
+import threading
+import urllib
+import queue
 
 from absl import app
 from absl import flags
 
+from rids import config
 from rids import network_capture
+from rids.iocs import iocs
+from rids import rules
 
 FLAGS = flags.FLAGS
 flags.DEFINE_string('host_ip', None,
                     'The IP of the host process, for ignoring direct requests')
-flags.DEFINE_string('config_path', 'config.json',
-                    'Path where IOC configuration can be found; uses config.json by default')
-
-
-def retrieve_bad_ips():
-  bad_ips = {}
-  with open(FLAGS.config_path, 'r') as f:
-    config = json.load(f)
-    for source in config['ioc_sources']:
-      # Download a fresh version of each IOC source in the config.
-      with urllib.request.urlopen(source['url']) as ioc_data:
-        # TODO check format of ioc source, branch to different parsers
-        # For now, the only source format is newline-seprated bad IPv4 addresses
-        # TODO perform parsing in a separate function from config handling
-        for line in ioc_data.readlines():
-          line = line.strip().decode('utf-8')
-          if not line: continue
-          bad_ip = str(ipaddress.ip_address(line))
-          if bad_ip not in bad_ips:
-            bad_ips[bad_ip] = [source['name']]
-          else:
-            bad_ips[bad_ip].append(source['name'])
-  return bad_ips
-
-
-def get_external_ip():
-  with urllib.request.urlopen('https://ipinfo.io/ip') as my_ip:
-    return my_ip.read().decode("utf-8").strip()
+flags.DEFINE_string('config_path', '/etc/rids/config.json',
+                    'file path indicating where IOC configuration can be found')
+flags.DEFINE_string('eventlog_path', '/var/rids/events.log',
+                    'file path indicating where to append new event logs')
 
 
 def main(argv):
+  """Main entry point of the app.  Also bound to CLI `rids` by setuptools."""
   if len(argv) > 1:
     raise app.UsageError('Too many command-line arguments.')
 
+  logging.basicConfig(level=logging.DEBUG,
+                      format='%(asctime)s %(levelname)-8s %(message)s',
+                      datefmt='%m-%d %H:%M',
+                      filename=FLAGS.eventlog_path,
+                      filemode='a')
+
+  cfg = config.ParseFile(FLAGS.config_path)
+  ruleset = rules.RuleSet()
+  for ioc_config in cfg['iocs']:
+    ruleset.MergeRuleset(iocs.parse_ruleset(ioc_config))
```

Comment:
I'd remove the MergeRuleset method. It's not necessary and restricts our development if we want to have other fields in the rule set (how do you merge them?).

Instead get rules from ioc_config and use AddRule()

Reply:
This conflicts with your suggestion elsewhere that favors AddIpRule(), AddConnectionRule() etc. which this main scope, and even the generalized iocs.parse_ruleset() scope, shouldn't need to know the kind of RuleSet.Add call to make.

I don't see how merge semantics restricts development.  The merging is meant to handle separate fields separately -- they wouldn't need to interfere.  Consider for example bad IPs vs bad domain names, it just needs to merge like with like.  It gets more interesting with e.g. IOCs of entire subnets being marked suspicious, as that could share a representation with the bad IPs but if so it would likely be indexed differently than the dict{} being used now for IPs.  Having separate indexes for specific IPs and subnets is just as suitable, anyway.

I was also thinking about the periodic re-fetching of IOC feeds.  By defining RuleSet as a per-feed and universal representation, a RuleSet could have all entries from a feed purged and updated with the new definitions, using the same process as used for fetching the initial feed definitions.  This can be done with a series of AddRule() or Add[type]Rule() calls too, though, so it doesn't really need Merge, it just seemed a suitable level of expression.

I abstracted this ruleset merging out of main(), removed the MergeRuleset() definition and I now have the IOC feed processing pass around a ruleset to the parsers like a context object.  This allows the feed-format specific processing call AddIpRule() etc. and main doesn't have to know about merging rulesets.